source/type/type.cpp

```diff
@@ -305,6 +306,56 @@ bool ValidityPattern::uncheck2(unsigned int day) const {
         return !days[day-1] && !days[day] && !days[day+1];
 }
 
+void MetaVehicleJourney::cancel_vj(RTLevel level,
+        const std::vector<boost::gregorian::date>& dates,
+        const transit_realtime::TripUpdate& /*trip_update*/,
+        const type::Data& data) {
+    auto func = [&](nt::VehicleJourney& vj){
+        for (const auto& date: dates){
+            // the train is canceled on one day, we just need to unset its realtime validitypattern
+            if (! vj.get_validity_pattern_at(level)->check(date)) { return; }
+            LOG4CPLUS_INFO(log4cplus::Logger::getInstance("realtime"),
+                    "canceling the vj " << vj.uri << " on " << date);
+            nt::ValidityPattern tmp_vp(*vj.get_validity_pattern_at(level));
+            tmp_vp.remove(date);
+            vj.validity_patterns[level] = data.pt_data->get_or_create_validity_pattern(tmp_vp);
+        }
+    };
+
+    for_vjs_at_rt_level(RTLevel::Base, func);
+    // we need to cancel all vj of at asked level
+    for_vjs_at_rt_level(level, func);
```

Comment:
No, the func must be called on every vjs. If you remove a realtime vj at a given date, but there is no rt vj this date, the adapted one will be the one valid, and we need to remove the date on it.